packages/ra-ui-materialui/src/list/filter/FilterListItem.tsx

```diff
@@ -144,27 +145,30 @@ const useStyles = makeStyles(theme => ({
  */
 const FilterListItem: FC<{ label: string; value: any }> = props => {
     const { label, value } = props;
-    const { filterValues, setFilters } = useListFilterContext();
+    const {
+        filterValues,
+        setFilters,
+        displayedFilters,
+        hideFilter,
+    } = useListContext();
     const translate = useTranslate();
     const classes = useStyles(props);
 
-    const isSelected = isEqual(value, filterValues);
+    const isSelected =
+        isEmpty(value) || isEmpty(filterValues)
+            ? false
+            : isMatch(filterValues, value);
 
     const addFilter = () => {
-        setFilters({ ...filterValues, ...value }, null, false);
+        setFilters(merge({}, filterValues, value), null, false);
     };
 
     const removeFilter = () => {
-        const keysToRemove = Object.keys(value);
-        const filters = Object.keys(filterValues).reduce(
-            (acc, key) =>
-                keysToRemove.includes(key)
-                    ? acc
-                    : { ...acc, [key]: filterValues[key] },
-            {}
-        );
-
-        setFilters(filters, null, false);
+        setFilters(omit(filterValues, paths(value)), null, false);
```

Comment:
Problem: we should update URL properly by removing only the leaf keys of value from filterValues.
Correction: using lodash omit and deepdash paths instead. Do you mind adding deepdash as a dependency ?